fuzz/fuzz_targets/serialization.rs

```diff
@@ -0,0 +1,209 @@
+#![no_main]
+use arbitrary::Arbitrary;
+use bson::{
+    raw::{RawBson, RawBsonRef, RawDocument},
+    spec::BinarySubtype,
+    Decimal128,
+};
+use libfuzzer_sys::fuzz_target;
+use std::str::FromStr;
+
+fn convert_bson_ref(bson_ref: RawBsonRef) -> Option<RawBson> {
+    match bson_ref {
+        RawBsonRef::Double(d) => {
+            if d.is_nan() {
+                Some(RawBsonRef::Double(f64::NAN).to_raw_bson())
+            } else if d.is_infinite() {
+                Some(RawBsonRef::Double(d).to_raw_bson())
+            } else {
+                Some(RawBsonRef::Double(d).to_raw_bson())
+            }
+        }
+        RawBsonRef::String(s) => {
+            if !s.is_empty() && !s.contains('\0') && s.len() <= (i32::MAX as usize) {
+                Some(RawBsonRef::String(s).to_raw_bson())
+            } else {
+                None
+            }
+        }
+        RawBsonRef::Document(d) => {
+            let mut valid = true;
+            for result in d.iter() {
+                match result {
+                    Ok((key, _)) if key.is_empty() || key.contains('\0') => {
+                        valid = false;
+                        break;
+                    }
+                    Err(_) => {
+                        valid = false;
+                        break;
+                    }
+                    _ => {}
+                }
+            }
+            if valid {
+                Some(RawBsonRef::Document(d).to_raw_bson())
+            } else {
+                None
+            }
+        }
+        RawBsonRef::Array(a) => {
+            let mut valid = true;
+            for result in a.into_iter() {
+                if result.is_err() {
+                    valid = false;
+                    break;
+                }
+            }
+            if valid {
+                Some(RawBsonRef::Array(a).to_raw_bson())
+            } else {
+                None
+            }
+        }
+        RawBsonRef::Binary(b) => {
+            if b.bytes.len() <= i32::MAX as usize &&
+               match b.subtype {
+                   BinarySubtype::Generic |
+                   BinarySubtype::Function |
+                   BinarySubtype::BinaryOld |
+                   BinarySubtype::UuidOld |
+                   BinarySubtype::Uuid |
+                   BinarySubtype::Md5 |
+                   BinarySubtype::UserDefined(_) => true,
+                   _ => false
+               } {
+                Some(RawBsonRef::Binary(b).to_raw_bson())
+            } else {
+                None
+            }
+        }
+        RawBsonRef::ObjectId(id) => Some(RawBsonRef::ObjectId(id).to_raw_bson()),
```

Comment:
This is interesting. I often don't like catch-alls because they can cause errors when new types are added, but I don't think the bson spec will be adding types any time soon.